triton-constraint-circuit/src/lib.rs

```diff
@@ -1471,4 +1705,316 @@ mod tests {
+                constraint.circuit.borrow().evaluate(
+                    main_input.view(),
+                    aux_input.view(),
+                    &challenges,
+                )
+            })
+            .collect_vec();
+
+        // apply one step of degree-lowering
+        let num_nodes = multicircuit_monad[0].builder.all_nodes.borrow().len();
+        let (mut substitution_node_id, mut substituted) = multicircuit_monad[0]
+            .builder
+            .all_nodes
+            .borrow()
+            .iter()
+            .nth(substitution_node_index % num_nodes)
+            .map(|(i, n)| (*i, n.clone()))
+            .unwrap();
+        let mut j = 0;
+        while substituted.circuit.borrow().is_zero() && j < num_nodes {
+            (substitution_node_id, substituted) = multicircuit_monad[0]
+                .builder
+                .all_nodes
+                .borrow()
+                .iter()
+                .nth((substitution_node_index + j) % num_nodes)
+                .map(|(i, n)| (*i, n.clone()))
+                .unwrap();
+            j += 1;
+        }
+        prop_assert_ne!(j, num_nodes, "no suitable nodes to substitute");
+
+        let degree_lowering_info = DegreeLoweringInfo {
+            target_degree: 2,
+            num_main_cols: main_input.len(),
+            num_aux_cols: aux_input.len(),
+        };
+        let substitution_constraint = ConstraintCircuitMonad::apply_substitution(
+            &mut multicircuit_monad,
+            degree_lowering_info,
+            substitution_node_id,
+            0,
+            0,
+        );
+
+        // extract substituted constraint
+        let CircuitExpression::BinOp(BinOp::Add, variable, neg_expression) =
+            &substitution_constraint.circuit.as_ref().borrow().expression
+        else {
+            unreachable!();
+        };
+        let extra_input =
+            match &neg_expression.as_ref().borrow().expression {
+                CircuitExpression::BinOp(BinOp::Mul, _neg_one, circuit) => circuit
+                    .borrow()
+                    .evaluate(main_input.view(), aux_input.view(), &challenges),
+                CircuitExpression::BConst(c) => -c.lift(),
+                CircuitExpression::XConst(c) => -*c,
+                _ => {
+                    unreachable!()
+                }
+            };
+        if variable.borrow().evaluates_to_base_element() {
+            main_input
+                .append(
+                    Axis(1),
+                    Array2::from_shape_vec([1, 1], vec![extra_input.coefficients[0]])
+                        .unwrap()
+                        .view(),
+                )
+                .unwrap();
+        } else {
+            aux_input
+                .append(
+                    Axis(1),
+                    Array2::from_shape_vec([1, 1], vec![extra_input])
+                        .unwrap()
+                        .view(),
+                )
+                .unwrap();
+        }
```

Comment:
I find the following more readable:

```suggestion
        if variable.borrow().evaluates_to_base_element() {
            let extra_input = extra_input.unlift().unwrap();
            let extra_input = Array2::from_shape_vec([1, 1], vec![extra_input]).unwrap();
            main_input.append(Axis(1), extra_input.view()).unwrap();
        } else {
            let extra_input = Array2::from_shape_vec([1, 1], vec![extra_input]).unwrap();
            aux_input.append(Axis(1), extra_input.view()).unwrap();
        }
```

Note that using `extra_input.unlift()` makes use of information hiding: it is an implementation detail how to go from `XFieldElement` to `BFieldElement`.

If you dislike the shadowing, I'm happy about other names. I think this is an acceptable case of variable name shadowing, because (a) the semantics of the variable does not change, and (b) because the shadowed variable gets incorporated into the new binding (what clippy refers to as [`shadow_reuse`](https://rust-lang.github.io/rust-clippy/master/index.html#/shadow_reuse), an allow-by-default lint[^allow]).

[^allow]: This part is a weak argument since `shadow_unrelated`, which I understand you dislike, is also allow-by-default.